packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,205 @@
+    for (const key of Object.keys(newFieldMarks)) {
+        const newMarkList = newFieldMarks[key];
+        baseFieldMarks[key] ??= [];
+        const baseMarkList = foldInMarkList(newMarkList, baseFieldMarks[key]);
+        baseFieldMarks[key] = baseMarkList;
+        while (typeof baseMarkList[baseMarkList.length - 1] === "number") {
+            baseMarkList.pop();
+        }
+        if (baseMarkList.length === 0) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+            delete baseFieldMarks[key];
+        }
+    }
+}
+
+function foldInMarkList(
+    newMarkList: T.MarkList,
+    baseMarkList: T.MarkList,
+): T.MarkList {
+    const factory = new MarkListFactory();
+    let iBase = 0;
+    let iNew = 0;
+    let nextBaseMark: T.Mark | undefined = baseMarkList[iBase];
+    let nextNewMark: T.Mark | undefined = newMarkList[iNew];
+    while (nextNewMark !== undefined) {
+        let newMark: T.Mark = nextNewMark;
+        let baseMark: T.Mark = nextBaseMark;
+        nextNewMark = undefined;
+        nextBaseMark = undefined;
+        if (baseMark === undefined) {
+            factory.push(clone(newMark));
+        } else if (isAttachGroup(newMark) || isReattach(newMark)) {
+            factory.pushContent(clone(newMark));
+            nextBaseMark = baseMark;
+        } else if (isDetachMark(baseMark)) {
+            // TODO: match base detaches to tombs and reattach in the newMarkList
+            factory.pushContent(baseMark);
+            nextNewMark = newMark;
+        } else {
+            const newMarkLength = getMarkLength(newMark);
+            const baseMarkLength = getMarkLength(baseMark);
+            if (newMarkLength < baseMarkLength) {
+                [baseMark, nextBaseMark] = splitMark(baseMark, newMarkLength);
+            } else if (newMarkLength > baseMarkLength) {
+                [newMark, nextNewMark] = splitMark(newMark, baseMarkLength);
+            }
+            // Passed this point, we are guaranteed that `newMark` and `baseMark` have the same length
+            if (typeof baseMark === "number") {
+                // TODO: insert new tombs and reattaches without replacing the offset
+                factory.push(newMark);
+            } else {
+                const composedMark = composeMarks(newMark, baseMark);
+                factory.push(...composedMark);
+            }
+        }
+        if (nextBaseMark === undefined) {
+            iBase += 1;
+            nextBaseMark = baseMarkList[iBase];
+        }
+        if (nextNewMark === undefined) {
+            iNew += 1;
+            nextNewMark = newMarkList[iNew];
+        }
+    }
+    if (nextBaseMark !== undefined) {
+        factory.push(nextBaseMark);
+    }
+    factory.push(...baseMarkList.slice(iBase + 1));
+    return factory.list;
+}
+
+function composeMarks(
+    newMark: T.SizedMark,
+    baseMark: T.ObjectMark | T.AttachGroup,
```

Comment:
It looks like AttachGroup is part of ObjectMark 

```suggestion
    baseMark: T.ObjectMark,
```